modules/guides/src/Controller/GuidesController.php

```diff
@@ -48,68 +64,66 @@ public static function create(ContainerInterface $container) {
    * @return array
    *   Render array.
    */
-  public function listGuides() {
-    $guides = [];
-    $guides_dir = DRUPAL_ROOT . (Settings::get('guides_dir') ?? '/guides');
-
-    foreach (array_diff(scandir($guides_dir), ['..', '.']) as $guide_dir) {
-      $dir = $guides_dir . '/' . $guide_dir;
-      if (is_dir($dir)) {
-        foreach (glob($dir . '/*.md') as $md) {
-          if (file_exists($md)) {
-            $parts = pathinfo($md);
-            $link = Url::fromRoute('guides.guide', ['filename' => $parts['filename']]);
-            $link = Link::fromTextAndUrl(str_replace('_', ' ', $guide_dir), $link)->toString();
-            $guides[] = [
-              '#markup' => $link,
-            ];
-          }
-        }
-      }
-    }
+  public function listGuides(): array {
+    // Get the links of the files.
+    $guides = $this->guidesStorage->getLinks();
 
+    // If the guides array isn't empty, return its content as item list.
     if (!empty($guides)) {
       return [
         '#theme' => 'item_list',
         '#items' => $guides,
       ];
     }
+    // Else return markup text.
     else {
       return [
-        '#markup' => $this->t('<strong>@text</strong>', ['@text' => 'No guides found.']),
+        '#type' => 'html_tag',
+        '#tag' => 'strong',
+        '#value' => $this->t('No guides found.'),
       ];
     }
   }
 
   /**
    * Page callback that renders a markdown file on the UI.
    *
+   * @param string $filename
+   *   The name of the guide file.
+   *
    * @return array
    *   Render array.
+   *
+   * @throws \Drupal\guides\Exception\ParseException
+   *   Thrown when no guide file is found.
    */
-  public function guideContent($filename) {
-    $guides_dir = Settings::get('guides_dir') ?? '/guides';
+  public function guideContent(string $filename): array {
     $target = [
       'dir' => FALSE,
       'file' => FALSE,
     ];
 
-    foreach (array_diff(scandir(DRUPAL_ROOT . $guides_dir), ['..', '.']) as $guide_dir) {
-      $dir = DRUPAL_ROOT . $guides_dir . '/' . $guide_dir;
-      $file = $dir . '/' . $filename . '.md';
-      if (file_exists($file)) {
-        $target['dir'] = $guides_dir . '/' . $guide_dir;
-        $target['file'] = $file;
+    // Loop through the directories inside the main guides directory.
+    foreach ($this->guidesStorage->getDirectories() as $guide_dir) {
+      // Put together the file path.
+      $file_path = $guide_dir . '/' . $filename . '.md';
+      // If the file exists with the given path then add dir and file keys to
+      // the target array.
+      if (file_exists($file_path)) {
+        $target['dir'] = $guide_dir;
+        $target['file'] = $file_path;
         break;
       }
     }
 
-    if (!$target['file']) {
-      throw new NotFoundHttpException();
+    try {
+      // Parses the guide file.
+      $md = new Parsedown();
+      $md = $md->text(str_replace('@guide_path', $target['dir'], file_get_contents($target['file'])));
```

Comment:
Instead of `file_get_contents()` and the loop over the `$this->guidesStorage->getDirectories()` lets add a new `getFileContent()` method to the guides_storage service that returns the content of a file _if exists_ or throws a module-specific FileNotFoundException. That exception does not need to be caught here because the route access control handler already validates that the file exists before this method gets called.

Reply:
Maybe this new method could also perform this `str_replace()` that is being done here.

Reply:
I'm not sure whether I solved this well or not, because I still had to use a loop to put together the path of the file from the file name.